akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala

```diff
@@ -0,0 +1,110 @@
+    majorityMinCap: Int)
+    extends Actor
+    with ActorLogging {
+
+  implicit val node: Cluster = Cluster(context.system)
+  implicit val selfUniqueAddress: SelfUniqueAddress = SelfUniqueAddress(node.selfUniqueAddress)
+
+  private val readMajority = ReadMajority(settings.tuningParameters.waitingForStateTimeout, majorityMinCap)
+  private val writeMajority = WriteMajority(settings.tuningParameters.updatingStateTimeout, majorityMinCap)
+
+  private val AllShardsKey = GSetKey[String](s"shard-${typeName}-all")
+  private var allShards: Option[Set[ShardId]] = None
+  private var coordinatorWaitingForShards: Option[ActorRef] = None
+
+  // eager load of remembered shard ids
+  def getAllShards(): Unit = {
+    replicator ! Replicator.Get(AllShardsKey, readMajority)
+  }
+  getAllShards()
+
+  override def receive: Receive = {
+    case RememberEntitiesCoordinatorStore.GetShards =>
+      allShards match {
+        case Some(shardIds) => sender() ! RememberEntitiesCoordinatorStore.RememberedShards(shardIds)
+        case None           =>
+          // reply when we get them, since there is only ever one coordinator communicating with us
+          // and it may retry we can just keep the latest sender
+          coordinatorWaitingForShards = Some(sender())
+      }
+
+    case g @ Replicator.GetSuccess(AllShardsKey, _) =>
+      onGotAllShards(g.get(AllShardsKey).elements)
+
+    case Replicator.NotFound(AllShardsKey, _) =>
+      onGotAllShards(Set.empty)
+
+    case Replicator.GetFailure(AllShardsKey, _) =>
+      log.error(
+        "The ShardCoordinator was unable to get all shards state within 'waiting-for-state-timeout': {} millis (retrying)",
+        readMajority.timeout.toMillis)
+      // repeat until GetSuccess
+      getAllShards()
+
+    case RememberEntitiesCoordinatorStore.AddShard(shardId) =>
+      replicator ! Replicator.Update(AllShardsKey, GSet.empty[String], writeMajority, Some((sender(), shardId)))(
+        _ + shardId)
+
+    case Replicator.UpdateSuccess(AllShardsKey, Some((replyTo: ActorRef, shardId: ShardId))) =>
+      log.debug("The coordinator shards state was successfully updated with {}", shardId)
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateDone(shardId)
+
+    case Replicator.UpdateTimeout(AllShardsKey, Some((replyTo: ActorRef, shardId: ShardId))) =>
+      log.error(
+        "The ShardCoordinator was unable to update shards distributed state within 'updating-state-timeout': {} millis (retrying), adding shard={}",
+        writeMajority.timeout.toMillis,
+        shardId)
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateFailed(shardId)
+
+    case Replicator.ModifyFailure(key, error, cause, Some((replyTo: ActorRef, shardId: ShardId))) =>
+      log.error(
+        cause,
+        "The remember entities store was unable to add shard [{}] (key [{}], failed with error: {})",
+        shardId,
+        key,
+        error)
+      replyTo ! RememberEntitiesCoordinatorStore.UpdateFailed(shardId)
```

Comment:
A bit odd we haven't different retry mechanisms for the two updates but i understand why

Reply:
`ModifyFailure` isn't something that will happen unless we have a bug. It's when the Update function throws.